node/basicnode/map_test.go

```diff
@@ -46,3 +51,277 @@ func BenchmarkSpec_Unmarshal_Map3StrInt(b *testing.B) {
+
+	expect := `map{
+	string{"cat"}: string{"meow"}
+	string{"dog"}: string{"bark"}
+	string{"eel"}: string{"zap"}
+}`
+	qt.Check(t, expect, qt.Equals, actual)
+
+	// copy the map using AssignNode
+	c := basicnode.Prototype.Map.NewBuilder()
+	err = c.AssignNode(mapNode)
+	if err != nil {
+		t.Fatal(err)
+	}
+	anotherNode := c.Build()
+
+	actual = printer.Sprint(anotherNode)
+	qt.Assert(t, expect, qt.Equals, actual)
+
+	// access values of map, using string
+	r, err := anotherNode.LookupByString("cat")
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "meow", qt.Equals, must.String(r))
+
+	// access values of map, using node
+	r, err = anotherNode.LookupByNode(basicnode.NewString("dog"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "bark", qt.Equals, must.String(r))
+
+	// access values of map, using PathSegment
+	r, err = anotherNode.LookupBySegment(datamodel.ParsePathSegment("eel"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "zap", qt.Equals, must.String(r))
+
+	// validate the node's prototype
+	np = anotherNode.Prototype()
+	qt.Check(t, fmt.Sprintf("%T", np), qt.Equals, "basicnode.Prototype__Map")
+}
+
+// test that AssignNode will fail if called twice, it expects an empty
+// node to assign to
+func TestMapCantAssignNodeTwice(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
+
+	// construct a map of three keys, using the MapBuilder
+	ma, err := b.BeginMap(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	a := ma.AssembleKey()
+	a.AssignString("cat")
+	a = ma.AssembleValue()
+	a.AssignString("meow")
+
+	a, err = ma.AssembleEntry("dog")
+	if err != nil {
+		t.Fatal(err)
+	}
+	a.AssignString("bark")
+
+	a = ma.AssembleKey()
+	a.AssignString("eel")
+	a = ma.AssembleValue()
+	a.AssignString("zap")
+
+	err = ma.Finish()
+	if err != nil {
+		t.Fatal(err)
+	}
+	mapNode := b.Build()
+
+	// copy the map using AssignNode, works the first time
+	c := basicnode.Prototype.Map.NewBuilder()
+	err = c.AssignNode(mapNode)
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Assert(t,
+		func() {
+			_ = c.AssignNode(mapNode)
+		},
+		qt.PanicMatches,
+		// TODO(dustmop): Error message here should be better
+		`misuse`)
+}
+
+func TestMapLookupError(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
+
+	// construct a map of three keys, using the MapBuilder
+	ma, err := b.BeginMap(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	a := ma.AssembleKey()
+	a.AssignString("cat")
+	a = ma.AssembleValue()
+	a.AssignString("meow")
+
+	a, err = ma.AssembleEntry("dog")
+	if err != nil {
+		t.Fatal(err)
+	}
+	a.AssignString("bark")
+
+	a = ma.AssembleKey()
+	a.AssignString("eel")
+	a = ma.AssembleValue()
+	a.AssignString("zap")
+
+	err = ma.Finish()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mapNode := b.Build()
+
+	_, err = mapNode.LookupByString("frog")
+	qt.Check(t, err, qt.ErrorMatches, `key not found: "frog"`)
+
+	_, err = mapNode.LookupByNode(basicnode.NewInt(3))
+	// TODO(dustmop): This error message is not great. It's about how the
```

Comment:
+1 you're right about this, feel free to open issues for things like this that are easy fixes - we can label them as "good-first-issue" or similar because we have folks looking for entry points to contribute. Or .. one of us could get to it eventually.

Reply:
Filed https://github.com/ipld/go-ipld-prime/issues/455